### PR TITLE
Load entry point models

### PR DESCRIPTION
**Pull request by Ivanidzo4ka**
Address https://github.com/dotnet/machinelearning/issues/1104
As long as NimbusML uses old version ml.net (pre 0.9) I don't have other way around other than this hacky way of loading it.

### CODE REVIEW

File: src/Microsoft.ML.Data/DataLoadSave/TransformerChain.cs [hunk 1]
```diff
@@ -252,8 +252,47 @@ public static TransformerChain<ITransformer> LoadFrom(IHostEnvironment env, Stre
...
-                ModelLoadContext.LoadModel<TransformerChain<ITransformer>, SignatureLoadModel>(env, out var transformerChain, rep, LoaderSignature);
-                return transformerChain;
+                try
+                {
+                    ModelLoadContext.LoadModel<TransformerChain<ITransformer>, SignatureLoadModel>(env, out var transformerChain, rep, LoaderSignature);
+                    return transformerChain;
+                }
+                catch
+                {
+                    var chain = ModelFileUtils.LoadPipeline(env, stream, new MultiFileSource(null), extractInnerPipe: false);
```
Comment (TomFinley) [inline src/Microsoft.ML.Data/DataLoadSave/TransformerChain.cs]: Usually when you do something that has a bad code smell (e.g., some sort of global `try/catch` of literally any type of exception ***whatsoever***) it is either best to either (1) explain why it must be so or (2) try not to do it. :) Either is fine with me.

File: src/Microsoft.ML.Data/DataView/RowToRowMapperTransform.cs [hunk 2]
```diff
@@ -44,6 +45,11 @@ internal interface IRowMapper : ICanSaveModel
         /// Returns information about the output columns, including their name, type and any metadata information.
         /// </summary>
         Schema.DetachedColumn[] GetOutputColumns();
+
+        /// <summary>
+        /// Returns parent transfomer which uses this mapper.
+        /// </summary>
+        ITransformer GetTransformer();
```
Comment (TomFinley) [inline src/Microsoft.ML.Data/DataView/RowToRowMapperTransform.cs]: So, this one strikes me as a little mysterious. The reason why `IRowMapper` exists *at all* is to make constructing a single `ITransformer` more convenient -- it is analogous in some respects, in this way, to a more serious `IValueMapper`, and a more specific mapper than `IRowToRowMapper`. So why does *this* specific thing need it? I suspect that it doesn't really belong here.

Comment (Ivanidzo4ka) [reply src/Microsoft.ML.Data/DataView/RowToRowMapperTransform.cs]: I'm just trying to achieve my goal of loading models in old format in less invasive way.
Right now I can load CompositeLoader which contains list of RowToRowMapperTransforms, (I don't even know is it Normalize, Pca, or any other transform, just RowToRowMapperTransform) if transform was converted, or just instance of IDataTransform.
If it's RowToRowMapperTransform, I need a way to get real ITransformer out of it, and mapper (which is IRowMapper) is only thing which I can use.

So basically since Mapper is pretty much all the time know it's parent (ITransformer) I'm just adding this method to it in order to get it back during walk through CompositeLoader transforms.



---
In reply to: [244916610](https://github.com/dotnet/machinelearning/pull/1951#discussion_r244916610) [](ancestors = 244916610)

Comment (TomFinley) [reply src/Microsoft.ML.Data/DataView/RowToRowMapperTransform.cs]: Sorry @ivanidzo4ka if I have missed something -- that's entirely possible -- but what does that have to do with this interface? All that tells me is that by using `CompositeDataLoader` we used the wrong class, or *it* is using the wrong class, or something. I mean, if I had a structure that took `IEnumerable`, but then I realized I wanted in fact the `IList` or whatever, I would think, "ok, this structure should take an `IList` instead!" I wouldn't think, "gee, can't possibly change the structure, I should change `IEnumerable` to produce the `IList` which it really is, since practically that's going to be the source of the `IEnumerable` all the time." I mean, that just seems backwards.

In addition to that, it seems contrary to our plans for how we've previously discussed how to load old models. The plan, as I recall, was the following: things that had been saved using the old (now disfavored) `IDataTransform` interface models, we will attempt to load as the new, shiny, `ITransformer` implementations. So: a normalizer (for instance) saved as `IDataTransform` in old version, we will read that form as an `ITransformer`. If it is something we have deliberately chosen *not* to convert to an `ITransform` (e.g., a filter), then it is dropped. (Pursuant to #993.) We would load them as `ITransformer` or nothing. We wouldn't have some "liminal" state where we use the old structures (which are being retired!) before converting them to the "real" structures we actually want. We would just use the new structures directly.

Now, from what I see, we are not doing that. Instead, I see we are *keeping* the `IDataTransform` instances and all the other baggage associated with them, and continue to rely on their presence. I am in fact going to keep using not only `IDataTransform`, but this old `IDataLoader` interface implementation `CompositeDataLoader`. But, because what we really wanted to do was load `ITransform` instances, we are now providing a bunch of backdoors through these created objects to recover the original `ITransform` instances upon which they are now based, instead of just using `ITransformer` directly and being done with it. I mean, why is `IDataTransform` or `IDataLoader` entering the picture *at all*? Doesn't that just complicate our ultimate desire to do away with them? (Pursuant to #1995.)

I get that you wanted to get away with just re-using the old `CompositeDataLoader`, and so had to make a bunch of old compromises because of that choice, but the correct thing to do was not use it, is it not so?

This seems architecturally a little backwards to me.

---
In reply to: [245146290](https://github.com/dotnet/machinelearning/pull/1951#discussion_r245146290) [](ancestors = 245146290,244916610)

Comment (TomFinley) [reply src/Microsoft.ML.Data/DataView/RowToRowMapperTransform.cs]: Ah OK, so this is intended to be a temporary situation. Could we somehow clarify this on `IRowMapper` interface? It just seems so odd, and the comment is not describing its "temporary nature" too well at the moment.

**Comment by Zruty0** [comment]
Is this work being abandoned? I viewed this as quite important, since Nimbus models were not loadable with ML.NET. Has that been updated?

File: src/Microsoft.ML.Data/DataView/RowToRowMapperTransform.cs (hunk 2 above)
Comment (Ivanidzo4ka) [reply src/Microsoft.ML.Data/DataView/RowToRowMapperTransform.cs]: I change description and modified comment to state what this is temporary solution.

Comment (sfilipi) [reply src/Microsoft.ML.Data/DataView/RowToRowMapperTransform.cs]: @ivmatan@microsoft.com the world makes more sense to me after reading this comment. Shall we save this explanation as code comments in the TransformerChain.LoadFrom?

---
In reply to: [245194301](https://github.com/dotnet/machinelearning/pull/1951#discussion_r245194301) [](ancestors = 245194301)

**Comment by sfilipi** [comment]
@ivmatan@microsoft.com i see that you added a bunch of models, in the test/data/backcompat . Did you intend to use them in tests?
